src/rgw/rgw_rados.cc

```diff
@@ -8769,14 +8795,39 @@ int RGWRados::cls_bucket_list_ordered(const DoutPrefixProvider *dpp,
     next_candidate(cct, tracker, candidates, tracker_idx);
 
     if (tracker.at_end() && tracker.is_truncated()) {
+      // Scenario: many sibling directories, many objects under each directory,
+      // if cls_filtered is true, tracker's entries may be the same as each other.
+      // the next_candidate operation may advance to end directly,
+      // so need to add candidates's entries which less than last key in the tracker, otherwise,
+      // only one object may be returned, this will slow down the list operation.
+      string end_key = tracker.retreat().entry_name();
+      auto citer = candidates.upper_bound(end_key);
+      for(auto iter = candidates.begin(); iter != citer; iter++) {
+        size_t tracker_idx = iter->second;
+        auto& tracker = results_trackers.at(tracker_idx);
+        const string& name = tracker.entry_name();
+        rgw_bucket_dir_entry& dirent = tracker.dir_entry();
+
+        r = handle_object(dpp, ioctx, tracker.oid_name, tracker.shard_idx, name, dirent, m, bucket_info,
+          updates, &count, &last_entry_visited, y, force_check_filter);
+        if (r < 0 && r != -ENOENT) {
+          return r;
+        }
+
+        ldpp_dout(dpp, 10) << "RGWRados::" << __func__ << " addtional key: "<< name <<
+          " added, shard " << tracker.oid_name << dendl;
+
+      }
+
       // once we exhaust one shard that is truncated, we need to stop,
       // as we cannot be certain that one of the next entries needs to
       // come from that shard; S3 and swift protocols allow returning
       // fewer than what was requested
       ldpp_dout(dpp, 10) << __PRETTY_FUNCTION__ <<
-	": stopped accumulating results at count=" << count <<
-	", dirent=\"" << dirent.key <<
-	"\", because its shard is untruncated and exhaused" << dendl;
+       ": stopped accumulating results at count=" << count <<
+       ", dirent=\"" << dirent.key <<
+       "\", because its shard: " << tracker.oid_name << " is untruncated and exhaused" << dendl;
```

Comment:
If you do change this, we can fix two issues with this message, which I think I originally wrote. First, it's spelled "exhausted", with a "t". Second, it should be "truncated", without the "un".